src/Mono.Android/Android.Runtime/JNIEnv.cs

```diff
@@ -254,12 +254,20 @@ public static unsafe IntPtr CreateInstance (Type type, string signature, params
 
 		public static IntPtr FindClass (System.Type type)
 		{
-			int rank = JavaNativeTypeManager.GetArrayInfo (type, out type);
 			try {
-				return FindClass (JavaNativeTypeManager.ToJniName (GetJniName (type), rank));
+				var sig  = JNIEnvInit.androidRuntime?.TypeManager.GetTypeSignature (type) ?? default;
+				if (!sig.IsValid || sig.SimpleReference == null) {
+					throw new ArgumentException ($"Could not determine Java type corresponding to `{type.AssemblyQualifiedName}`.", nameof (type));
+				}
+
+				JniObjectReference local_ref = JniEnvironment.Types.FindClass (sig.Name);
+				IntPtr global_ref = NewGlobalRef (local_ref.Handle);
+				JniObjectReference.Dispose (ref local_ref);
+				return global_ref;
 			} catch (Java.Lang.Throwable e) {
 				if (!((e is Java.Lang.NoClassDefFoundError) || (e is Java.Lang.ClassNotFoundException)))
 					throw;
+				int rank = JavaNativeTypeManager.GetArrayInfo (type, out type);
```

Comment:
We should log or something in the error path so we have something to look for in NativeAOT builds to see if we're hitting the error path.

Reply:
It calls `monodroid_log` for `e` currently, which would also fail spectacularly...

I put on the list to cleanup `monodroid_log()` along with the Android system property parsing:
* https://github.com/dotnet/android/issues/9784